File: LeetCode/Problems/Python/218. The Skyline Problem.py
#Heap + sort
#Time: O(nlogn)
#Space: O(n)
class Solution(object):
    def getSkyline(self, buildings):
        
        #we want to sort by height so put H in front
        #we want to separate L and R, so set height of R as 0
        #since if it's the end, it will be (R,0)
        critical_pts = [[L, -H, R] for L, R, H in buildings]
        critical_pts.extend([[R, 0, 0] for L, R, H in buildings])
        critical_pts.sort()
        
        #we can use a heap to track the current alive building during the scan from left to right
        #so we initial the result and alive list with default value
        #result:[x, height]
        #initial with one item for more general algo
        result = [[0,0]]
        
        #heap 
        #[-height, ending position]
        alive = [(0, float('inf'))]        
        
        
        for pos, negH, R in critical_pts:
            #step 1: check if any building aready end?
            while alive[0][1] <= pos:
                heapq.heappop(alive)
            #step 2: if the current pt starts a new building
            if negH < 0:
                heapq.heappush(alive, (negH, R))
                #print(alive)

            #step 3: check if we need to add new pt to result
            #        we can see if the height change, because it indicates
            #        1. new building 
            #        2. we've passed old building's R
            if result[-1][1] != -alive[0][0]:
                result.append([pos, -alive[0][0]])
            
        return result[1:]

File: LeetCode/Problems/C++/218. The Skyline Problem.cpp
class Solution {
public:
    vector<vector<int>> getSkyline(vector<vector<int>>& buildings) {
        if(!buildings.size()) return {};
        
        vector<vector<int>> critical_pt;
        
        //use long for height
        priority_queue<pair<int, long>> alive;
        vector<vector<int>> result;
        for(int i=0; i<buildings.size();i++){
            critical_pt.push_back({buildings[i][0], -buildings[i][2], buildings[i][1]});
            critical_pt.push_back({buildings[i][1], 0, 0});
        }
        sort(critical_pt.begin(), critical_pt.end());
        
        //height, right
        alive.push({0, LONG_MAX});

        
        for(int i=0; i<critical_pt.size();i++){
            //check if buildings r are smaller than i
            while(alive.top().second <= critical_pt[i][0]){
                alive.pop();
            }
            
            //check if i is the start of the new building
            if(critical_pt[i][1]!=0){
                alive.push({-critical_pt[i][1], critical_pt[i][2]});
            }
            
            //check if height change, and add to result
            if(i==0 || (result.back()[1] != alive.top().first)){
                result.push_back({critical_pt[i][0], alive.top().first});
            }
        }

        return result;
        
    }
};